spec/cypress/e2e/submissions_spec.cy.js

```diff
@@ -1,26 +1,124 @@
-      FactoryBot.create("assignment", { lecture_id: 1 }).as("assignment");
-      cy.createUser("admin").as("user");
+    // 🐰 "Joiner" joins the submission using the code
+    cy.login(this.joiner).then(() => {
+      subscribeToLecture(this.lecture.id);
+      joinSubmissionViaToken(this.lecture.id, this.token);
+      cy.logout();
+    });
+
+    // New assignment
+    Timecop.moveAheadDays(1000).then(() => {
+      FactoryBot.create("assignment", { lecture_id: this.lecture.id }).as("assignment");
+
+      // 🎈🎈 "Inviter2" creates a submission & stores code
+      cy.login(this.inviter2).then(() => {
+        subscribeToLecture(this.lecture.id);
+        createEmptySubmission(this.lecture.id);
+        cy.logout();
+      });
+
+      // 🐰 "Joiner" joins the submission using the code
+      cy.login(this.joiner).then(() => {
+        joinSubmissionViaToken(this.lecture.id, this.token);
+        cy.logout();
+      });
+    });
+
+    // New assignment
+    Timecop.moveAheadDays(2000).then(() => {
+      FactoryBot.create("assignment", { lecture_id: this.lecture.id }).as("assignment");
+
+      // 🎈 "Inviter" invites "Joiner" to a new submission
+      cy.login(this.inviter).then(() => {
+        // The "Joiner" name is not prefilled here since for the last assignment
+        // "Inviter" did not submit anything (only "Inviter2" did)
+        createEmptySubmission(this.lecture.id, this.joiner.name_in_tutorials);
+        cy.logout();
+      });
 
-      cy.then(() => {
-        console.log(this.tutorial);
-        console.log(this.assignment);
-        console.log(this.user);
+      // 🎈🎈 "Inviter2" invites "Joiner" to a new submission
+      cy.login(this.inviter2).then(() => {
+        createEmptySubmission(this.lecture.id);
+        cy.logout();
       });
 
-      cy.then(() => {
-        cy.visit("/lectures/1/");
+      // 🐰 "Joiner" can now join without a code
+      // (since this user has previously handed in a submission together
+      // with the inviter, see above)
+      cy.login(this.joiner).then(() => {
+        cy.visit(`/lectures/${this.tutorial.lecture_id}/submissions`);
+        cy.getBySelector("submission-join").click();
+        cy.getBySelector("accept-invite-0").click();
+        cy.getBySelector("submission-team").should("contain", this.inviter.name_in_tutorials);
+        cy.getBySelector("submission-leave").click();
+
+        // now there is only one invite left (that from "Inviter2")
+        cy.getBySelector("submission-join").click();
+        cy.getBySelector("accept-invite-1").should("not.exist");
+        cy.getBySelector("accept-invite-0").should("be.visible");
+        cy.getBySelector("submission-cancel-join").click();
+
+        // can also join on parent page
+        cy.getBySelector("accept-invite-0").click();
+        cy.getBySelector("submission-team").should("contain", this.inviter2.name_in_tutorials);
       });
     });
   });
```

Comment:
You are using `name_in_tutorials` here and some lines above. This is a database field that might be `nil` (of course, it is not in the tests). The method `tutorial_name` in the user model provides a fallback (if `name_in_tutorials` is nil it takes the `name` field which can't be nil). Of course, that does not change anything in the tests, but it might be a good idea to use `tutorial_name` here as well just as in `submission_invite.rb` for consistency. But I am not insisting on that so feel free to resolve if you don't this is necessary.

Reply:
Unfortunately, this is highly non-trivial to implement since `tutorial_name` is a function, and not a property. While for objects returned by `FactoryBot` I've already implemented a dynamic call mechanism, the user object is not a FactoryBot object and thus needs special treatment. Such a feature would go far beyond the scope of this PR.

Proposal: for now, let's use the private field in the test. Should we encounter this scenario more often, we can think about implementing the dynamic call mechanism for user objects as well.

Reply:
I agree.